### PR TITLE
[FSDP][Perf] Do not call `pad` in no-padding case

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #88453 [Dynamo][FSDP] Migrate to `ModuleWrapPolicy`
* #88450 [FSDP] Introduce `ModuleWrapPolicy` for simplicity
* **#88769 [FSDP][Perf] Do not call `pad` in no-padding case**

- Calling `F.pad()` issues a pad kernel from the CPU even if there is no padding needed, which can incur some non-negligible overhead. This PR removes that unnecessary call for the no-padding case.
- This PR also does not zero the newly-allocated sharded gradient tensor before the reduce-scatter if `use_orig_params=True` because there is no need. The reduce-scatter will fill the tensor anyway, and we do not care about the values in the padding. For `use_orig_params=False`, the padding is exposed to the user, so we preserve the existing semantics of zeroing it. I left a to-do to follow-up since we may optimize that.

### CODE REVIEW

**Review by zhaojuanmao — APPROVED** [review]

**Comment by awgu** [comment]
`F.pad(tensor, [0, 0])` actually copies the tensor.
```
>>> import torch
>>> t = torch.randn((3, 3))
>>> t.data_ptr()
93936708252800
>>> torch.nn.functional.pad(t, [0, 0]).data_ptr()
93936640304064
```
This allocation happened previously in the post-backward stream, which induced cross-stream memory fragmentation. (Only the _sharded_ gradient needs to be allocated in the post-backward stream, not the _unsharded_ gradient.)

For T5-11B on 2 nodes and batch size 6, eliminating the unnecessary `F.pad` for the no-padding case decreases peak reserved memory by ~300 MB and leads to 0 CUDA malloc retries instead of 3.
